Assets/PackageStagingArea/eu.netherlands3d.twin-functionalities/Scripts/Functionality.cs

```diff
@@ -1,13 +1,43 @@
-﻿using SimpleJSON;
+﻿using System;
+using SimpleJSON;
 using UnityEngine;
 using UnityEngine.Events;
 
 namespace Netherlands3D.Twin.Functionalities
 {
+    [Serializable]
+    public class FunctionalityData
+    {
+        public string Id;
+        public bool IsEnabled;
+    }
+
     [CreateAssetMenu(menuName = "Netherlands3D/Twin/Functionality", fileName = "Functionality", order = 0)]
     public class Functionality : ScriptableObject, ISimpleJsonMapper
     {
-        public string Id;
+        [SerializeField] private FunctionalityData data = new();
+        public FunctionalityData Data
+        {
+            get
+            {
+                return data;
+            }
+            set
+            {
+                var oldEnabled = data.IsEnabled;
+                data = value;
+                
+                //invoke events if the state of the new Data object and old Data object don't match
+                if (data.IsEnabled != oldEnabled)
```

Comment:
Ik zie hier duplicatie tov de property hier beneden; is het praktisch om dat naar -in ieder geval een private- setter methode te verplaatsen zodat we niet de duplicatie hoeven te tracken?

Reply:
Ik denk het niet, wat ik bedoel is dat ik dit stukje code op 2 plaatsen zie: in de Data property setter, en in de IsEnabled property setter. Dit is duplicatie van hetzelfde concept; en daarmee gevoelig voor bugs waarbij maar 1 van de 2 wordt aangepast bij toekomstige wijzigingen.

Mijn suggestie zou zijn om dat stukje code in een private methode onder te brengen (zodat het in dit Data object blijft) en beide plekken diezelfde methode te laten aanroepen. Dan heb je toch dat op 2 plekken hetzelfde gebeurd; maar dan zonder de duplicatie

Reply:
ah zo, nee dat is niet precies hetzelfde, als je het Data object set, dan moet je de isEnabled waarde van de nieuwe data vergelijken met de isEnabled van de oude. Het kan denk ik wel versimpeld worden door eerst de waarde te setten van de oude Data en daarme de events invoken, en daarna het data object kan overschrijven. De enige issue is dan als deze events dan het Data object opvragen van de functionality, dat gebeurt dan voordat data = newObject wordt gedaan.